cloudinit/distros/networking.py

```diff
@@ -185,17 +186,51 @@ def try_set_link_up(self, devname: DeviceName) -> bool:
 class BSDNetworking(Networking):
     """Implementation of networking functionality shared across BSDs."""
 
+    def __init__(self):
+        self.ifc = ifconfig.Ifconfig()
+        self.ifs = {}
+        self._update_ifs()
+
+    def _update_ifs(self):
+        ifconf = subp.subp(["ifconfig", "-a"])
+        # ``ifconfig -a`` always returns at least ``lo0``.
+        # So this ``if`` is really just to make testing/mocking easier
+        if ifconf[0]:
+            self.ifs = self.ifc.parse(ifconf[0])
+
     def apply_network_config_names(self, netcfg: NetworkConfig) -> None:
         LOG.debug("Cannot rename network interface.")
 
     def is_physical(self, devname: DeviceName) -> bool:
-        raise NotImplementedError()
+        return self.ifs[devname].is_physical
+
+    def is_bond(self, devname: DeviceName) -> bool:
+        return self.ifs[devname].is_bond
+
+    def is_bridge(self, devname: DeviceName) -> bool:
+        return self.ifs[devname].is_bridge
+
+    def is_vlan(self, devname: DeviceName) -> bool:
+        return self.ifs[devname].is_vlan
+
+    def is_up(self, devname: DeviceName) -> bool:
+        return self.ifs[devname].up
 
     def settle(self, *, exists=None) -> None:
         """BSD has no equivalent to `udevadm settle`; noop."""
 
     def try_set_link_up(self, devname: DeviceName) -> bool:
-        raise NotImplementedError()
+        """Try setting the link to up explicitly and return if it is up.
+        Not guaranteed to bring the interface up. The caller is expected to
+        add wait times before retrying."""
+        subp.subp(["ifconfig", devname, "up"])
+        return self.is_up(devname)
+
+    def get_devicelist(self) -> list:
+        return list(self.ifc.ifs_by_name().keys())
+
+    def get_interface_mac(self, devname: DeviceName):
+        return self.ifs[devname].mac
```

Comment:
Weirdly, this isn't being called:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 767, in status_wrapper
    ret = functor(name, args)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 433, in main_init
    init.apply_network_config(bring_up=bring_up_interfaces)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/stages.py", line 924, in apply_network_config
    self.distro.networking.wait_for_physdevs(netcfg)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/distros/networking.py", line 149, in wait_for_physdevs
    present_macs = self.get_interfaces_by_mac().keys()
  File "/usr/local/lib/python3.9/site-packages/cloudinit/distros/networking.py", line 77, in get_interfaces_by_mac
    blacklist_drivers=self.blacklist_drivers
AttributeError: 'FreeBSDNetworking' object has no attribute 'blacklist_drivers'
```

it looks like what is being called is the one from `Networking`

Reply:
it looks like the most sensible approach is to backout this change, and do a proper refactor for everything that's not an `@abc.abstractmethod` yet in `Networking`.

Reply:
done.